File: genesis/create-genesis.js
const fs = require('fs');
const eth = require('ethereumjs-util');
const BigNumber = require('bignumber.js');
const utils = require('web3-utils');

const GENESIS_CONFIG = {
  "config": {
    "chainId": 0,
    "homesteadBlock": 0,
    "eip150Block": 0,
    "eip150Hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "eip155Block": 0,
    "eip158Block": 0,
    "byzantiumBlock": 0,
    "constantinopleBlock": 0,
    "petersburgBlock": 0,
    "istanbulBlock": 0,
    "muirGlacierBlock": 0,
    "parlia": {
      "period": 12,
      "epoch": 10,
    }
  },
  "nonce": "0x0",
  "timestamp": "0x5e9da7ce",
  "extraData": "0x00",
  "gasLimit": "0x2625a00",
  "difficulty": "0x1",
  "mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "coinbase": "0x0000000000000000000000000000000000000000",
  "alloc": {
    "0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE": {
      "balance": "0x0"
    },
    "0x86d274133714A88CE821F279e5eD3fb0BfB42503": {
      "balance": "0x21e19e0c9bab2400000"
    },
    "0x00a601f45688dba8a070722073b015277cf36725": {
      "balance": "0x21e19e0c9bab2400000"
    },
    "0xbAdCab1E02FB68dDD8BBB0A45Cc23aBb60e174C8": {
      "balance": "0x21e19e0c9bab2400000"
    },
    "0x57BA24bE2cF17400f37dB3566e839bfA6A2d018a": {
      "balance": "0x21e19e0c9bab2400000"
    },
    "0xEbCf9D06cf9333706E61213F17A795B2F7c55F1b": {
      "balance": "0x21e19e0c9bab2400000"
    },
  },
  "number": "0x0",
  "gasUsed": "0x0",
  "parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000"
};

const DEPLOYER_ADDRESS = "0x0000000000000000000000000000000000000010";
const GOVERNANCE_ADDRESS = "0x0000000000000000000000000000000000000020";
const PARLIA_ADDRESS = "0x0000000000000000000000000000000000000030";

const SYSTEM_CONTRACTS = [
  {
    address: DEPLOYER_ADDRESS,
    artifactPath: "./build/contracts/DeployerV1.json",
  },
  {
    address: GOVERNANCE_ADDRESS,
    artifactPath: "./build/contracts/GovernanceV1.json",
  },
  {
    address: PARLIA_ADDRESS,
    artifactPath: "./build/contracts/ParliaV1.json",
  },
];

const createExtraData = (validators) => {
  const zeroVanity = '0'.repeat(64);
  const validatorList = validators.map(v => {
    return v.startsWith('0x') ? v.substr(2) : v
  }).join('')
  const zeroSeal = '0'.repeat(130);
  return `0x${zeroVanity}${validatorList}${zeroSeal}`
};

const createGenesisConfig = (chainId, validators) => {
  const genesisConfig = Object.assign({}, GENESIS_CONFIG);
  for (const {address, artifactPath} of SYSTEM_CONTRACTS) {
    const {deployedBytecode} = require(artifactPath);
    const balance = {
      balance: '0x00',
      code: deployedBytecode,
    };
    // if (address === PARLIA_ADDRESS) {
    //   const PARLIA_VALIDATOR_MAP_OFFSET = 50,
    //     PARLIA_VALIDATOR_ARRAY_OFFSET = 51;
    //   const storage = {};
    //   validators.forEach((v, i) => {
    //     // fill mapping
    //     const mappingOffset = eth.toBuffer(utils.padLeft(v, 64)),
    //       mappingKey = eth.toBuffer(utils.padLeft(PARLIA_VALIDATOR_MAP_OFFSET, 64))
    //     const mappingHash = eth.keccak256(Buffer.concat([mappingOffset, mappingKey])).toString('hex')
    //     storage[`0x${mappingHash}`] = `0x000000000000000000000001${v.substr(2)}`;
    //     // fill array
    //     const arrayKeyOffset = eth.keccak256(eth.toBuffer(utils.padLeft(PARLIA_VALIDATOR_ARRAY_OFFSET, 64))),
    //       arrayIndexKey = utils.padLeft(new BigNumber(arrayKeyOffset.toString('hex'), 16).plus(i).toString(16), 64)
    //     storage[`0x${arrayIndexKey}`] = utils.padLeft(v, 64);
    //   });
    //   const arraySizeHash = eth.toBuffer(utils.padLeft(PARLIA_VALIDATOR_ARRAY_OFFSET, 64))
    //   storage[`0x${arraySizeHash.toString('hex')}`] = utils.padLeft(validators.length, 64);
    //   balance.storage = storage;
    // }
    genesisConfig.alloc[address] = balance;
  }
  genesisConfig['config']['chainId'] = chainId;
  genesisConfig['extraData'] = createExtraData(validators);
  return genesisConfig;
};

(async () => {
  const genesisMainnet = createGenesisConfig(17242, []);
  const genesisTestnet = createGenesisConfig(17243, [
    '0x00A601f45688DbA8a070722073B015277cF36725',
  ]);
  fs.writeFileSync('./mainnet/genesis.json', JSON.stringify(genesisMainnet, null, 2));
  fs.writeFileSync('./testnet/genesis.json', JSON.stringify(genesisTestnet, null, 2));
})();

File: genesis/create-genesis.go
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"io/ioutil"
	"math/big"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/consensus"
	"github.com/ethereum/go-ethereum/core"
	"github.com/ethereum/go-ethereum/core/rawdb"
	"github.com/ethereum/go-ethereum/core/state"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/core/vm"
	"github.com/ethereum/go-ethereum/ethdb/memorydb"
	"github.com/ethereum/go-ethereum/params"
	"github.com/ethereum/go-ethereum/trie"
)

type artifactData struct {
	Bytecode         string `json:"bytecode"`
	DeployedBytecode string `json:"deployedBytecode"`
}

func createExtraData(validators []common.Address) []byte {
	extra := make([]byte, 32+20*len(validators)+65)
	for i, v := range validators {
		copy(extra[32+20*i:], v.Bytes())
	}
	return extra
}

type dummyChainContext struct {
}

func (d *dummyChainContext) Engine() consensus.Engine {
	return nil
}

func (d *dummyChainContext) GetHeader(h common.Hash, n uint64) *types.Header {
	return nil
}

var deployerAddress = common.HexToAddress("0x0000000000000000000000000000000000000010")
var governanceAddress = common.HexToAddress("0x0000000000000000000000000000000000000020")
var parliaAddress = common.HexToAddress("0x0000000000000000000000000000000000000030")
var systemAddress = common.HexToAddress("0xfffffffffffffffffffffffffffffffffffffffe")

func simulateSystemContract(genesis *core.Genesis, systemContract common.Address, rawArtifact []byte, constructor []byte) error {
	artifact := &artifactData{}
	if err := json.Unmarshal(rawArtifact, artifact); err != nil {
		return err
	}
	bytecode := append(hexutil.MustDecode(artifact.Bytecode), constructor...)
	// simulate constructor execution
	ethdb := rawdb.NewDatabase(memorydb.New())
	db := state.NewDatabaseWithConfig(ethdb, &trie.Config{})
	statedb, err := state.New(common.Hash{}, db, nil)
	if err != nil {
		return err
	}
	block := genesis.ToBlock(nil)
	blockContext := core.NewEVMBlockContext(block.Header(), &dummyChainContext{}, &common.Address{})
	txContext := core.NewEVMTxContext(
		types.NewMessage(common.Address{}, &systemContract, 0, big.NewInt(0), 10_000_000, big.NewInt(0), []byte{}, nil, false),
	)
	evm := vm.NewEVM(blockContext, txContext, statedb, genesis.Config, vm.Config{})
	deployedBytecode, _, err := evm.CreateWithAddress(vm.AccountRef(common.Address{}), bytecode, 10_000_000, big.NewInt(0), systemContract)
	if err != nil {
		return err
	}
	contractState := statedb.GetOrNewStateObject(systemContract)
	storage := contractState.GetDirtyStorage()
	// read state changes from state database
	genesisAccount := core.GenesisAccount{
		Code:    deployedBytecode,
		Storage: storage,
		Balance: big.NewInt(0),
		Nonce:   0,
	}
	fmt.Printf("Affected storage for contract: %s\n", systemContract.Hex())
	for key, value := range storage {
		fmt.Printf(" ~ %s -> %s\n", key.Hex(), value.Hex())
	}
	if genesis.Alloc == nil {
		genesis.Alloc = make(core.GenesisAlloc)
	}
	genesis.Alloc[systemContract] = genesisAccount
	return nil
}

//go:embed build/contracts/Deployer.json
var deployerRawArtifact []byte

//go:embed build/contracts/Governance.json
var governanceRawArtifact []byte

//go:embed build/contracts/Parlia.json
var parliaRawArtifact []byte

func newArguments(typeNames ...string) abi.Arguments {
	var args abi.Arguments
	for i, tn := range typeNames {
		abiType, err := abi.NewType(tn, tn, nil)
		if err != nil {
			panic(err)
		}
		args = append(args, abi.Argument{Name: fmt.Sprintf("%d", i), Type: abiType})
	}
	return args
}

type genesisConfig struct {
	Genesis         *core.Genesis
	Deployers       []common.Address
	Validators      []common.Address
	GovernanceOwner common.Address
	VotingPeriod    int64
	Faucet          map[common.Address]string
}

func createGenesisConfig(config genesisConfig, targetFile string) error {
	genesis := config.Genesis
	// extra data
	genesis.ExtraData = createExtraData(config.Validators)
	// execute system contracts
	ctor, err := newArguments("address[]").Pack(config.Deployers)
	if err != nil {
		return err
	}
	if err := simulateSystemContract(genesis, deployerAddress, deployerRawArtifact, ctor); err != nil {
		return err
	}
	ctor, err = newArguments("address", "uint256").Pack(config.GovernanceOwner, big.NewInt(config.VotingPeriod))
	if err != nil {
		return err
	}
	if err := simulateSystemContract(genesis, governanceAddress, governanceRawArtifact, ctor); err != nil {
		return err
	}
	ctor, err = newArguments("address[]").Pack(config.Validators)
	if err != nil {
		return err
	}
	if err := simulateSystemContract(genesis, parliaAddress, parliaRawArtifact, ctor); err != nil {
		return err
	}
	// create system contract
	genesis.Alloc[systemAddress] = core.GenesisAccount{
		Balance: big.NewInt(0),
	}
	// apply faucet
	for key, value := range config.Faucet {
		balance, ok := new(big.Int).SetString(value[2:], 16)
		if !ok {
			return fmt.Errorf("failed to parse number (%s)", value)
		}
		genesis.Alloc[key] = core.GenesisAccount{
			Balance: balance,
		}
	}
	// save to file
	newJson, _ := json.MarshalIndent(genesis, "", "  ")
	return ioutil.WriteFile(targetFile, newJson, fs.ModePerm)
}

func defaultGenesisConfig(chainId int64) *core.Genesis {
	chainConfig := &params.ChainConfig{
		ChainID:             big.NewInt(chainId),
		HomesteadBlock:      big.NewInt(0),
		EIP150Block:         big.NewInt(0),
		EIP155Block:         big.NewInt(0),
		EIP158Block:         big.NewInt(0),
		ByzantiumBlock:      big.NewInt(0),
		ConstantinopleBlock: big.NewInt(0),
		PetersburgBlock:     big.NewInt(0),
		IstanbulBlock:       big.NewInt(0),
		MuirGlacierBlock:    big.NewInt(0),
		RamanujanBlock:      big.NewInt(0),
		NielsBlock:          big.NewInt(0),
		MirrorSyncBlock:     big.NewInt(0),
		BrunoBlock:          big.NewInt(0),
		Parlia: &params.ParliaConfig{
			Period: 3,
			Epoch:  200,
		},
	}
	return &core.Genesis{
		Config:     chainConfig,
		Nonce:      0,
		Timestamp:  0x5e9da7ce,
		ExtraData:  nil,
		GasLimit:   0x2625a00,
		Difficulty: big.NewInt(0x01),
		Mixhash:    common.Hash{},
		Coinbase:   common.Address{},
		Alloc:      nil,
		Number:     0x00,
		GasUsed:    0x00,
		ParentHash: common.Hash{},
	}
}

var testnetConfig = genesisConfig{
	Genesis: defaultGenesisConfig(17242),
	// who is able to deploy smart contract from genesis block
	Deployers: []common.Address{
		common.HexToAddress("0x00a601f45688dba8a070722073b015277cf36725"),
		common.HexToAddress("0xbAdCab1E02FB68dDD8BBB0A45Cc23aBb60e174C8"),
		common.HexToAddress("0xEbCf9D06cf9333706E61213F17A795B2F7c55F1b"),
		common.HexToAddress("0x57BA24bE2cF17400f37dB3566e839bfA6A2d018a"),
	},
	// list of default validators
	Validators: []common.Address{
		common.HexToAddress("0x00a601f45688dba8a070722073b015277cf36725"),
	},
	// owner of the governance
	GovernanceOwner: common.HexToAddress("0x00a601f45688dba8a070722073b015277cf36725"),
	VotingPeriod:    60, // 3 minutes
	// faucet
	Faucet: map[common.Address]string{
		common.HexToAddress("0x86d274133714A88CE821F279e5eD3fb0BfB42503"): "0x21e19e0c9bab2400000",
		common.HexToAddress("0x00a601f45688dba8a070722073b015277cf36725"): "0x21e19e0c9bab2400000",
		common.HexToAddress("0xbAdCab1E02FB68dDD8BBB0A45Cc23aBb60e174C8"): "0x21e19e0c9bab2400000",
		common.HexToAddress("0x57BA24bE2cF17400f37dB3566e839bfA6A2d018a"): "0x21e19e0c9bab2400000",
		common.HexToAddress("0xEbCf9D06cf9333706E61213F17A795B2F7c55F1b"): "0x21e19e0c9bab2400000",
	},
}

func main() {
	if err := createGenesisConfig(testnetConfig, "testnet.json"); err != nil {
		panic(err)
	}
}
